R/content.R

```diff
@@ -1,6 +1,47 @@
 #' @include utils-S7.R
 NULL
 
+#' Format contents into a textual representation.
+#' 
+#' @description
+#' These generic functions can be use to convert [Turn] contents or [Content]
+#' objects into textual representations. 
+#' 
+#' * `contents_text()` is the most minimal and only includes [ContentText]
+#'   objects in the output.
+#' * `contents_markdown()` returns the text content (which it assumes to be
+#'   markdown and does not convert it) plus markdown representations of images
+#'   and other content types.
+#' * `contents_html()` returns the text content, converted from markdown to
+#'   HTML with [commonmark::markdown_html()], plus HTML representations of
+#'   images and other content types.
+#' 
+#' @examples
+#' turns <- list(
+#'   user_turn("What's this image?", content_image_url("https://placehold.co/200x200")),
+#'   Turn("assistant", "It's a placeholder image.")
+#' )
+#'
+#' lapply(turns, contents_text)
+#' lapply(turns, contents_markdown)
+#' if (rlang::is_installed("commonmark")) {
+#'   contents_html(turns[[1]])
+#' }
+#' 
+#' @param content The [Turn] or [Content] object to be converted into text.
```

Comment:
Can you pass turns?

Reply:
I just added a method for `contents_markdown()` for `Chat` instances. I see a lot of utility in that particular output format and less utility in the HTML or text variants. That said I'd be happy to add those if you want.

``` r
turns <- list(
  user_turn("What's this image?", content_image_url("https://placehold.co/200x200")),
  Turn("assistant", "It's a placeholder image.")
)

chat <- Chat$new(Provider("https://example.com"), turns = turns)
cat(contents_markdown(chat))
#> ## User
#> 
#> What's this image?
#> 
#> ![](https://placehold.co/200x200)
#> 
#> ## Assistant
#> 
#> It's a placeholder image.
```

Reply:
Related: is there a way to create a `chat_echo()` (or maybe `chat_repeater()`) chat instance that just returns the user input and doesn't call any APIs? Carson made an [example like this for Shiny](https://shiny.posit.co/py/components/display-messages/chat/) and I've found it to be very useful for debugging and examples when I'm not ready for real model usage yet.

Reply:
I think that would be pretty straightforward with a new provider.